### ISSUE TITLE
[BUG] Odd behavior of `cython.dataclasses.dataclass()` decorator

### ISSUE THREAD

**Issue by Bluenix2** [cython/cython#4624]
## Summary

- Extension dataclasses (`cdef dataclass`) in Pure Python mode has to be ordered with the `cython.cclass` decorator inner-most to the class.
- There's a compiler crash if `cython.dataclasses.dataclass()` is called with no arguments

## Description

The following code fails with `Decorators not allowed on cdef classes (used on type 'MyTest')`:
```python
import cython

@cython.cclass
@cython.dataclasses.dataclass
class MyTest:
    val: cython.int
```

The compiler crashes with an `AttributeError` on `NoneType` with the following code:
```python
import cython

@cython.cclass
@cython.dataclasses.dataclass()
class MyTest:
    val: cython.int
```
```python
ModuleNode.body = StatListNode(_cython_magic_00b187774452a423c4a80249cd19a8bb448a9ad2.pyx:2:0)
StatListNode.stats[1] = StatListNode(_cython_magic_00b187774452a423c4a80249cd19a8bb448a9ad2.pyx:6:0)
StatListNode.stats[0] = PyClassDefNode(_cython_magic_00b187774452a423c4a80249cd19a8bb448a9ad2.pyx:6:0,
    name = 'MyTest')

Compiler crash traceback from this point on:
  File "Cython\\Compiler\\Visitor.py", line 182, in Cython.Compiler.Visitor.TreeVisitor._visit
  File "[c:\Users\%username%\Projects\wumpy\venv\lib\site-packages\Cython\Compiler\ParseTreeTransforms.py]()", line 1229, in visit_PyClassDefNode
    directives = self._extract_directives(node, 'class')
  File "[c:\Users\%username%\Projects\wumpy\venv\lib\site-packages\Cython\Compiler\ParseTreeTransforms.py]()", line 1241, in _extract_directives
  File "[c:\Users\%username%\Projects\wumpy\venv\lib\site-packages\Cython\Compiler\ParseTreeTransforms.py]()", line 1101, in try_to_parse_directives
    directives.append(self.try_to_parse_directive(optname, args, kwds, node.function.pos))
  File "[c:\Users\%username%\Projects\wumpy\venv\lib\site-packages\Cython\Compiler\ParseTreeTransforms.py]()", line 1180, in try_to_parse_directive
    return (optname, (args, kwds.as_python_dict()))
AttributeError: 'NoneType' object has no attribute 'as_python_dict'
```

## Environment

- Windows 10
- Python 3.9.5 (in a `venv`)
- Cython pulled from `master` (can confirm that I have 2d287a6191bbba1e120ba9240b24ce364c443204)

**Comment by da-woods** [comment]
I think might apply to the other class decorators too (e.g. `cython.final`, `cython.autopickle`, `no_gc_clear`). Not completely sure though, but I think I've seen this before with other combinations.

**Comment by scoder** [comment]
> Extension dataclasses (`cdef dataclass`) in Pure Python mode has to be ordered with the `cython.cclass` decorator inner-most to the class.

I think it makes sense to require that. `@cclass` on a dataclass is different from `@dataclass` on a cdef class. Obviously shouldn't crash but give an appropriate error message.

**Comment by Bluenix2** [comment]
> > Extension dataclasses (`cdef dataclass`) in Pure Python mode has to be ordered with the `cython.cclass` decorator inner-most to the class.
> 
> I think it makes sense to require that. `@cclass` on a dataclass is different from `@dataclass` on a cdef class. Obviously shouldn't crash but give an appropriate error message.

I disagree. I don't think we should make much sense of it really. For the sake of providing an argument against your point though:

> The former, to me, makes more sense. The only way to make `@dataclass` work is to overwrite different methods of the class and doing so will create methods that use arbitrary attributes. After a class has been turned into an extension type it can no longer have arbitrary attributes, therefore you should be adding `@cclass` on a dataclass.
> 
> Besides, if we ever needed to have this make sense logically, then `cclass` would've been a metaclass instead of a class decorator.

The way I see it, `@cython.cclass` is just a marker that can be placed anywhere.

It would be unnecessary to restrict this and all we'd be doing is disallow code that visually looks better ordered that way.

**Comment by scoder** [comment]
Decorators are always ordered. Why allow something that just lets users write incorrect code? Being strict here gives us more freedom in the future, without any drawbacks for users.

**Comment by Bluenix2** [comment]
> Decorators are always ordered. Why allow something that just lets users write incorrect code? Being strict here gives us more freedom in the future, without any drawbacks for users.

So in this case you consider the correct code to be having `@cython.cclass` closest to the class, as if `cdef` was moved to the line above yeah?

In that light I do agree, although I generally think about it differently and mostly prefer the look with longer decorators closer to the class (when there's such a big difference) unless of course, there's a strictly logical reason they're ordered in the way they are (which is often not the case from the libraries I work with).

**Comment by da-woods** [comment]
> So in this case you consider the correct code to be having @cython.cclass closest to the class, as if cdef was moved to the line above yeah?

That's definitely how I'd look at it - it should be made a class first, then made a cython dataclass after that.

If we're disallowing it, I do think we should have a good clear error message though (e.g. "`cython.cclass` should be the innermost decorator") so at least the nobody wastes too much time trying to think about the fix.

**Comment by da-woods** [comment]
>  this one because there's no real logic in the decorators

I'd argue that there's actually quite a bit of logic in the `dataclass` decorator (both Cython and Python). As an example, imagine you have another decorator that does something to `__annotations__` or maybe adds an `__init__` function. The order would make a big difference to what Python's `dataclass` does

**Comment by Bluenix2** [comment]
Although I haven't changed my preference; I have no more arguments for my cause and I accept defeat 😅. I think we can move forward with this.

I might look into it this weekend and see what I can do (but to be clear: it's up-for-grabs if anyone else gets to it before i do).

**Comment by Bluenix2** [comment]
So I did have a look at this and found the `_extract_directives()` method but did not manage to get anywhere else other than adding some prints to `try_to_parse_directives()` and figure out how `as_cython_attribute` works. I'll make it clear that I am no longer working on this.

**Comment by scoder** [comment]
> > Extension dataclasses (`cdef dataclass`) in Pure Python mode has to be ordered with the `cython.cclass` decorator inner-most to the class.
> 
> I think it makes sense to require that. `@cclass` on a dataclass is different from `@dataclass` on a cdef class. Obviously shouldn't crash but give an appropriate error message.

I revised my opinion on this. It is easy to forget `@cython.cclass` when using `@cython.dataclasses.dataclass`, and difficult to notice, because it keeps working the same way, just with a Python dataclass.

`@cclass` should be implied for `@cython.dataclasses.dataclass`.

We can still require `@cclass` for Python's own `@dataclass` decorator to distinguish between Python dataclasses and Cython dataclasses. In fact, Cython's separate decorator seems redundant, given that we can make the distinction based on `@cclass` being used together with Python's `@dataclass` decorator or not. I think that's a much simpler way to distinguish between Python dataclasses and extension type dataclasses.

**Comment by da-woods** [comment]
> It is easy to forget `@cython.cclass` when using `@cython.dataclasses.dataclass`, and difficult to notice, because it keeps working the same way, just with a Python dataclass.

I think it just does nothing rather than making it a Python dataclass. But either way it's easy to miss and the current behaviour isn't helpful.

> In fact, Cython's separate decorator seems redundant, given that we can make the distinction based on `@cclass` being used together with Python's `@dataclass` decorator or not.

I'm slightly undecided on this. Although on balance I think it's probably the best thing to do.
* In favour it does seem a lot more straightforward. I struggle to remember what needs importing from `@cython.dataclasses.*` and what should be imported from `dataclasses`, and I really should know.
* Against: I'm slightly uncomfortable with "taking over" someone else's module and imposing our own behaviour on it.
